engine/storage/src/main/java/org/apache/cloudstack/storage/image/deployasis/DeployAsIsHelperImpl.java

```diff
@@ -174,8 +174,25 @@ private void handleGuestOsFromOVFDescriptor(long templateId, String guestOsType,
                 hypervisor.toString(), minimumHypervisorVersion);
 
         if (CollectionUtils.isNotEmpty(guestOsMappings)) {
-            GuestOSHypervisorVO mapping = guestOsMappings.get(0);
-            long guestOsId = mapping.getGuestOsId();
+            Long guestOsId = null;
+            if (guestOsMappings.size() == 1) {
+                GuestOSHypervisorVO mapping = guestOsMappings.get(0);
+                guestOsId = mapping.getGuestOsId();
+            } else {
+                if (!StringUtils.isEmpty(guestOsDescription)) {
+                    for (GuestOSHypervisorVO guestOSHypervisorVO : guestOsMappings) {
+                        GuestOSVO guestOSVO = guestOSDao.findById(guestOSHypervisorVO.getGuestOsId());
+                        if (guestOsDescription.equalsIgnoreCase(guestOSVO.getDisplayName())) {
+                            guestOsId = guestOSHypervisorVO.getGuestOsId();
+                            break;
+                        }
+                    }
+                }
+                if (null == guestOsId) {
```

Comment:
Minor nit - I think variable == null maybe more readable, but as I said that's not picking

Reply:
depends on how accustomed you are in right to left reading ;)

Reply:
making habit of this to avoid setting null to the variable by mistake, in that case IDE also don't know if it is logically wrong "if (guestOsId = null)" :p